ansible_rulebook/builtin.py

```diff
@@ -610,6 +611,117 @@ async def post_process_runner(
+    set_facts: Optional[bool] = None,
+    post_events: Optional[bool] = None,
+    verbosity: int = 0,
+    var_root: Union[str, Dict, None] = None,
+    copy_files: Optional[bool] = False,
+    json_mode: Optional[bool] = False,
+    retries: Optional[int] = 0,
+    retry: Optional[bool] = False,
+    delay: Optional[int] = 0,
+    **kwargs,
+):
+
+    logger.info(
+        "running job template: %s, organization: %s", name, organization
+    )
+    logger.info("ruleset: %s, rule %s", source_ruleset_name, source_rule_name)
+
+    hosts_limit = ",".join(hosts)
+    if not job_args:
+        job_args = {}
+    job_args["limit"] = hosts_limit
+    job_id = str(uuid.uuid4())
+
+    await event_log.put(
+        dict(
+            type="Job",
+            job_id=job_id,
+            ansible_rulebook_id=settings.identifier,
+            name=name,
+            ruleset=source_ruleset_name,
+            rule=source_rule_name,
+            hosts=hosts_limit,
+            action="run_job_template",
+        )
+    )
+
+    async def event_callback(event: dict) -> None:
+        await event_log.put(
+            {
+                "type": "AnsibleEvent",
+                "event": {
+                    "uuid": event["uuid"],
+                    "counter": event["counter"],
+                    "stdout": event["stdout"],
+                    "start_line": event["start_line"],
+                    "end_line": event["end_line"],
+                    "event": event["event"],
+                    "created": event["created"],
+                    "parent_uuid": event["parent_uuid"],
+                    "event_data": event["event_data"],
+                    "job_id": job_id,
+                    "controller_job_id": event["job"],
+                    "ansible_rulebook_id": settings.identifier,
+                },
+                "run_at": event["created"],
+            }
+        )
+
+    if retry:
+        retries = max(retries, 1)
+    for i in range(retries + 1):
+        if i > 0:
+            if delay > 0:
+                await asyncio.sleep(delay)
+            logger.info(
+                "Previous run_job_template failed. Retry %d of %d", i, retries
+            )
+        controller_job = await job_template_runner.run_job_template(
+            name, organization, job_args, event_callback
+        )
+        if controller_job["status"] != "failed":
+            break
+
+    await event_log.put(
+        dict(
+            type="Action",
+            action="run_job_template",
+            activation_id=settings.identifier,
+            job_template_name=name,
+            organization=organization,
+            job_id=job_id,
+            ruleset=ruleset,
+            rule=source_rule_name,
+            status=controller_job["status"],
+            run_at=controller_job["created"],
+        )
+    )
+
+    if set_facts or post_events:
+        logger.debug("set_facts")
+        facts = controller_job["artifacts"]
```

Comment:
@bzwei is the artifacts always present should we check if its none before posting

Reply:
artifacts alway present, but could be `{}`.

Reply:
@bzwei so can we check we only post non empty artifacts